Add login form help button

bergice opened the pull request:
Issue https://github.com/silverstripe/silverstripe-framework/issues/9883

Before:
![image](https://user-images.githubusercontent.com/541886/112915406-b3873980-915a-11eb-95b7-272d22d77949.png)
After:
![image](https://user-images.githubusercontent.com/541886/112915233-6014eb80-915a-11eb-8752-189c3e75e179.png)

bergice committed:
Better describe the 'keep me signed in' option

bergice commented:
Resolves https://github.com/silverstripe/silverstripe-framework/issues/9883

bergice committed:
Use new designs

emteknetnz commented on themes/login-forms/templates/Security.ss:
```diff
@@ -11,14 +11,24 @@
         <meta name="color-scheme" content="light dark" />
         <% require css("silverstripe/admin: client/dist/styles/bundle.css") %>
         <% require css("silverstripe/login-forms: client/dist/styles/bundle.css") %>
+        <% require javascript('silverstripe/admin: thirdparty/jquery/jquery.js') %>
```
Including javascript from the admin thirdparty dir, which is intended for the CMS, not the frontend, is a bad idea.  It's likely that this will break in the future without us realising.  We generally don''t like the dependencies in the thirdparty dir and they don't recieve security updates and we'll refactor them out to be installed via npm/yarn instead.  Recent example of this https://github.com/silverstripe/silverstripe-userforms/pull/1045

You'll need to change this to get jquery / tooltip, etc, to install via yarn and then included in the frontend js bundle

emteknetnz replied:
Including javascript from the admin thirdparty dir, which is intended for the CMS, not the frontend, is a bad idea. It's likely that this will break in the future without us realising. We generally don''t like the dependencies in the thirdparty dir and they don't recieve security updates and we'll refactor them out to be installed via npm/yarn instead. Recent example of this silverstripe/silverstripe-userforms#1045

You'll need to change this to get jquery / tooltip, etc, to install via yarn and then included in a frontend js bundle

emteknetnz commented on themes/login-forms/templates/Security.ss:
```diff
@@ -11,14 +11,24 @@
+        <% require javascript('silverstripe/admin: thirdparty/popper/popper.min.js') %>
+        <% require javascript('silverstripe/admin: thirdparty/bootstrap/js/dist/util.js') %>
+        <% require javascript('silverstripe/admin: thirdparty/bootstrap/js/dist/tooltip.js') %>
+        <% require customScript("window.jQuery('body').tooltip({ selector: '[data-toggle=tooltip]' });", 'bootstrap.tooltip-boot') %>
```
We should avoid inlining scripts on the frontend where possible since that's usually the first thing that gets disabled off when [Content security policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) gets added.  When can't control what projects CSP implementation, so if any projects have a CSP policy with inline scripts disabled, this script simply won't run.

Instead move this to a js bundle

bergice replied:
Does this also affect admin? `vendor/silverstripe/admin/code/LeftAndMain.php:692`

emteknetnz replied:
I think if you added a strict CSP to admin now then things probably wouldn't work.

Sites implementing a CSP are likely to either do it via an .htaccess file in the public folder (which doesn't touch admin) or via a 3rd party module that would presumably also exclude admin.

emteknetnz commented on themes/login-forms/templates/SilverStripe/Forms/CheckboxField_holder.ss:
```diff
@@ -0,0 +1,13 @@
+        <% if $RightTitle %> $RightTitle<% end_if %>
+    </label>
+    <% if $getAttribute('title') && $getAttribute('data-toggle') == 'tooltip' %>
+    <i class="font-icon-help-circled" title="$getAttribute('title')" data-toggle="$getAttribute('data-toggle')"
```
Needs indentation

emteknetnz commented on themes/login-forms/templates/SilverStripe/Forms/CheckboxField_holder.ss:
```diff
@@ -0,0 +1,13 @@
+    </label>
+    <% if $getAttribute('title') && $getAttribute('data-toggle') == 'tooltip' %>
+    <i class="font-icon-help-circled" title="$getAttribute('title')" data-toggle="$getAttribute('data-toggle')"
+    ></i>
```
Remove line break so tag isn't split

emteknetnz commented on themes/login-forms/templates/Security.ss:
```diff
@@ -11,14 +11,24 @@
 
         <main class="login-form">
+            <div class="login-form__header">
             <% if $Title %>
```
Indentation is off

emteknetnz commented on client/src/styles/login-form.scss:
```diff
@@ -176,3 +186,12 @@ input[type="submit"] {
 }
+
+.field {
+  .font-icon-help-circled {
```
Can remove a nesting level and just
`.field .font-icon-help-circled { ... }`

bergice replied:
I think this is more readable.

emteknetnz commented:
Some other things:

a) Travis build failures - you'll need to update .travis.yml to use https://github.com/silverstripe/silverstripe-travis-shared - the `standard-jobs-range.yml` provision should be fine - just about every other repo has this so plenty of examples to copy from

b) Question mark icon need `cursor: pointer` css

c) The same tooltip is on BOTH the checkbox AND the question mark icon - this seems very weird.  It's probably supposed to only be on the question mark icon and not the checkbox.

d) Text alignment issues (firefox)

![image](https://user-images.githubusercontent.com/4809037/114107259-bd602800-9924-11eb-8ef4-017f5374aca7.png)

emteknetnz reviewed (changes requested):
Some other things:

a) Travis build failures - you'll need to update .travis.yml to use https://github.com/silverstripe/silverstripe-travis-shared - the `standard-jobs-range.yml` provision should be fine - just about every other repo has this so plenty of examples to copy from

b) Question mark icon need `cursor: pointer` css

c) The same tooltip is on BOTH the checkbox AND the question mark icon - this seems very weird.  It's probably supposed to only be on the question mark icon and not the checkbox. @clarkepaul ?

d) Question mark icon cannot be tabbed to

e) Text alignment issue with checkbox label (firefox)

![image](https://user-images.githubusercontent.com/4809037/114108267-d8cc3280-9926-11eb-80b5-21deedd051c5.png)

bergice committed:
Move bootstrap to npm

clarkepaul commented:
C. Yes only want it on the question mark, no underline on hover for the icon either. 

D. Would really want to get tabbing to work if possible. There should be a way we can get that to work... I know I wasn't able to in my code example I provided but think you could write better JS to accommodate it.

@bergice Mentioned elsewhere... for this amount of multi line content the popover component is more appropriate.

bergice committed:
Fix checkbox field rendering tooltip twice

bergice commented:
> b) Question mark icon need `cursor: pointer` css

Not sure if I agree with that, the icon is not clickable and the tooltip displays immediately anyways when hovered.

> c) The same tooltip is on BOTH the checkbox AND the question mark icon - this seems very weird. It's probably supposed to only be on the question mark icon and not the checkbox. @clarkepaul ?

Fixed.

> e) Text alignment issue with checkbox label (firefox)

This is already an issue from before:

![image](https://user-images.githubusercontent.com/541886/114466839-1984be00-9c3d-11eb-8cf7-f803136a23df.png)

But I've fixed it.

emteknetnz commented:
> This is already an issue from before

I'm confused by your screenshot - it's from the framework login form - this is the login-forms pull-request

bergice committed:
Fix checkbox alignment

bergice commented:
> > This is already an issue from before
> 
> I'm confused by your screenshot - it's from the framework login form - this is the login-forms pull-request

Pretty sure it applies to both, but I've fixed it for login-forms.

bergice committed:
• Move attribute to login-forms

emteknetnz commented on themes/login-forms/templates/SilverStripe/Forms/CheckboxField_holder.ss:
```diff
@@ -0,0 +1,12 @@
+        $Title
+        <% if $RightTitle %> $RightTitle<% end_if %>
+    </label>
+    <% if $getAttribute('title') && $getAttribute('data-toggle') == 'tooltip' %>
```
```suggestion
    <% if $getAttribute('title') %>
```

emteknetnz commented on themes/login-forms/templates/SilverStripe/Forms/CheckboxField.ss:
```diff
@@ -0,0 +1 @@
+<input $getAttributesHTML('title', 'data-toggle') />
```
I'm not even sure what this does - was this for enabling the tooltip on the checkbox?  We don't want to do this?  Can we just delete this file entirely?

bergice replied:
It's for doing the opposite. Those attributes are "exclusions", not "inclusions".

emteknetnz replied:
OK sweet

emteknetnz commented on client/src/js/bundle.js:
```diff
@@ -0,0 +1,4 @@
+import 'bootstrap/js/dist/util';
+import 'bootstrap/js/dist/tooltip';
+
+window.jQuery('body').tooltip({selector: '[data-toggle=tooltip]'});
```
not ideal - though just leave as is - we can update this with follow up pr here https://github.com/creative-commoners/silverstripe-login-forms/pull/1/files#diff-6fb8efb7c9bd3ed9c20610108631b6751c85ef17ac1cf0ed7e31bda800ed3c32R5

bergice replied:
How would you like to see this done?
This is how we do it in `admin` through a `customJavascript` call.

emteknetnz replied:
No need to update - I'm currently working through this here https://github.com/silverstripe/silverstripe-login-forms/pull/89/files#diff-6fb8efb7c9bd3ed9c20610108631b6751c85ef17ac1cf0ed7e31bda800ed3c32R5

emteknetnz reviewed (approved):
I've noted some changes we should make - though as agreed we can implement on follow up PR here https://github.com/creative-commoners/silverstripe-login-forms/pull/1

bergice committed:
Display tooltip when title is set

emteknetnz commented:
Screenshot of final

![image](https://user-images.githubusercontent.com/4809037/114625258-143f7600-9d06-11eb-9555-6c754f7690f3.png)
